File: src/main/kotlin/MoveHelper.java
import model.Game;
import model.Move;
import model.Wizard;
import model.World;

import java.util.List;

import static java.lang.StrictMath.*;

public class MoveHelper {

    private Wizard self;
    private World world;
    private Game game;
    private Move move;

    private Point2D lastFindingPoint = null;
    private Point2D lastFoundPoint = null;

    public MoveHelper(Wizard self, World world, Game game, Move move) {
        this.self = self;
        this.world = world;
        this.game = game;
        this.move = move;
    }

    public void goTo(Point2D point) {
        Point2D correctedPoint = correctPoint(point);

        double angle = self.getAngleTo(correctedPoint.getX(), correctedPoint.getY());
        move.setTurn(angle);

        goWithoutTurn(point);
    }

    public void goWithoutTurn(Point2D point) {
        Point2D correctedPoint = correctPoint(point);

        double diffAngle = self.getAngleTo(correctedPoint.getX(), correctedPoint.getY());

        double backCoef = cos(diffAngle);
        double strickCoef = sin(diffAngle);

        if (abs(diffAngle) > PI / 2) {
            move.setSpeed(game.getWizardBackwardSpeed() * backCoef);
        } else {
            move.setSpeed(game.getWizardForwardSpeed() * backCoef);
        }
        move.setStrafeSpeed(game.getWizardStrafeSpeed() * strickCoef);
    }

    public Point2D correctPoint(Point2D point2D) {
        if (point2D.equals(lastFindingPoint)) {
            if (lastFoundPoint != null) return lastFoundPoint;
            else return point2D;
        }

        lastFindingPoint = point2D;

        WayFinder wayFinder = new WayFinder(self, world, game);
        List<Point2D> way = wayFinder.findWay(point2D);

        if (way != null && way.size() > 0) {
            Point2D findPoint = way.get(0);
            lastFoundPoint = findPoint;

            return findPoint;
        }

        return point2D;
    }


}

File: src/main/kotlin/MoveHelper.kt
import model.*

import java.lang.StrictMath.*

class MoveHelper(private val self: Wizard, private val world: World, private val game: Game, private val move: Move) {

    fun goTo(unit: LivingUnit, increaseDistanceToMinion: Boolean = false) {
        goTo(unit.toPoint(), increaseDistanceToMinion)
    }

    fun goTo(point: Point2D, increaseDistanceToMinion: Boolean = false) {
        val correctedPoint = correctPoint(point, increaseDistanceToMinion)

        val angle = self.getAngleTo(correctedPoint.x, correctedPoint.y)
        move.turn = angle

        goWithoutTurn(point)
    }

    fun goWithoutTurn(point: Point2D) {
        val correctedPoint = correctPoint(point)

        val diffAngle = self.getAngleTo(correctedPoint.x, correctedPoint.y)

        val backCoef = cos(diffAngle)
        val strickCoef = sin(diffAngle)

        if (abs(diffAngle) > PI / 2) {
            move.speed = game.wizardBackwardSpeed * backCoef
        } else {
            move.speed = game.wizardForwardSpeed * backCoef
        }
        move.strafeSpeed = game.wizardStrafeSpeed * strickCoef
    }

    fun correctPoint(point2D: Point2D, increaseDistanceToMinion: Boolean = false): Point2D {
        val wayFinder = WayFinder(self, world, game)
        val way = wayFinder.findWay(point2D, increaseDistanceToMinion)

        if (way != null) {
            val findPoint = way

            return findPoint
        }

        return point2D
    }


}
